unixfs/src/dir/builder.rs

```diff
@@ -0,0 +1,132 @@
+use cid::Cid;
+use std::fmt;
+
+mod dir_builder;
+use dir_builder::DirBuilder;
+
+mod iter;
+pub use iter::{OwnedTreeNode, PostOrderIterator, TreeNode};
+
+mod buffered;
+pub use buffered::BufferingTreeBuilder;
+
+enum Entry {
+    Leaf(Leaf),
+    Directory(DirBuilder),
+}
+
+impl fmt::Debug for Entry {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Entry::*;
+
+        match self {
+            Leaf(leaf) => write!(fmt, "Leaf {{ {:?} }}", leaf),
+            Directory(_) => write!(fmt, "DirBuilder {{ .. }}"),
+        }
+    }
+}
+
+impl Entry {
+    fn as_dir_builder(&mut self) -> Result<&mut DirBuilder, ()> {
+        use Entry::*;
+        match self {
+            Directory(ref mut d) => Ok(d),
+            _ => Err(()),
+        }
+    }
+}
+
+struct Leaf {
+    link: Cid,
+    total_size: u64,
+}
+
+impl fmt::Debug for Leaf {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "{}, {}", self.link, self.total_size)
+    }
+}
+
+/// Configuration for customizing how the tree is built.
+#[derive(Default, Debug, Clone)]
+pub struct TreeOptions {
+    wrap_with_directory: bool,
+}
+
+impl TreeOptions {
+    /// When true, allow multiple top level entries, otherwise error on the second entry.
+    /// Defaults to false.
+    pub fn wrap_with_directory(&mut self) {
+        self.wrap_with_directory = true;
+    }
+}
+
+/// Tree building failure cases.
+#[derive(Debug)]
+pub enum TreeBuildingFailed {
+    /// The given full path started with a slash; paths in the `/add` convention are not rooted.
+    RootedPath(String),
+    /// The given full path contained empty segment.
+    RepeatSlashesInPath(String),
+    /// If the `BufferingTreeBuilder` was created without `TreeOptions` with the option `wrap in
+    /// directory` enabled, then there can be only a single element at the root.
+    TooManyRootLevelEntries,
+    /// The given full path had already been added.
+    DuplicatePath(String),
+    /// The given full path had already been added as a link to an opaque entry.
+    LeafAsDirectory(String),
+}
+
+impl fmt::Display for TreeBuildingFailed {
```

Comment:
I keep forgetting that there's https://crates.io/crates/thiserror (that we're already using) that could make this more concise, but even if it's desirable I wouldn't apply this on top of this PR; it would make a good `FirstIssue` though.

Reply:
I have mixed feelings on it. I get that it saves a few lines but it's just simple code. I'd much rather write the simple code than wonder what it's expanded, then try to find a correct item for `cargo expand`, get frustrated and expand the whole project and search from there.

Reply:
That totally depends on our approach to the crate in general (I have no strong feelings one way or the other); if we don't like it then a good `FirstIssue` would be to get rid of it :stuck_out_tongue:.

Reply:
just confirmed: there is no thiserror on the `cargo tree` at the moment.